pkg/controller/vitessshard/reconcile_disk.go

```diff
@@ -0,0 +1,126 @@
+package vitessshard
+
+import (
+	"context"
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+	apitypes "k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	planetscalev2 "planetscale.dev/vitess-operator/pkg/apis/planetscale/v2"
+	"planetscale.dev/vitess-operator/pkg/operator/results"
+	"planetscale.dev/vitess-operator/pkg/operator/rollout"
+)
+
+const (
+	pvcDiskSizeAnnotation = "pvc-disk-size"
+)
+
+func (r *ReconcileVitessShard) reconcileDisk(ctx context.Context, vts *planetscalev2.VitessShard) (reconcile.Result, error) {
+	resultBuilder := &results.Builder{}
+	changed := false
+	tabletKeys := tabletKeysFromShard(vts)
+	tabletPods, err := r.tabletPodsFromShard(ctx, vts)
+	if err != nil {
+		return resultBuilder.Error(err)
+	}
+
+	for i  := range vts.Spec.TabletPools {
+		tabletPool := &vts.Spec.TabletPools[i]
+		requestedDiskQuantity := tabletPool.DataVolumeClaimTemplate.Resources.Requests[v1.ResourceStorage]
+		requestedDisk := requestedDiskQuantity.String()
+		poolTablets := tabletsForPool(vts, tabletKeys, tabletPool.Cell, string(tabletPool.Type))
+
+		for _, tabletKey := range poolTablets {
+			pod, ok := tabletPods[tabletKey]
+			if !ok {
+				continue
+			}
+
+			tabletDiskSize, ok := pod.Annotations[pvcDiskSizeAnnotation]
+			if !ok {
+				// If it's never been set before, apply the disk size annotation.
+				setDiskSizeAnnotation(pod, requestedDisk)
+				continue
+			}
+
+			// If disk size is unchanged, move on to the next tablet.
+			if tabletDiskSize == requestedDisk {
```

Comment:
We can't look at the actual annotation on the Pod to determine whether it's time to cascade. The annotation won't be there until after the Pod is restarted; until then, it is merely a scheduled change.

What we want to check for here is, among all the PVCs whose Status size differs from the desired size configured in the tablet pool:

1. Do all those PVCs have their Spec size set to the new desired size?
2. Do all those PVCs have the status condition `FileSystemResizePending` set to `True`?
3. Do all the corresponding Pods have some kind of update scheduled (`rollout.Scheduled(pod)`)?

If all these are true for all tablet pools with such pending PVCs, and if the number of pending PVCs is greater than zero, then we should trigger a cascade (if one is not already ongoing).

Reply:
Done.

Reply:
As a note, I switched the order of checks 1 and 2 so that the logic with the anythingChanged variable stays the same, let me know if that breaks anything.